MekHQ/src/mekhq/campaign/personnel/education/EducationController.java

```diff
@@ -0,0 +1,1879 @@
+import java.util.*;
+
+/**
+ * The EducationController class is responsible for managing the education process.
+ * It provides methods to begin the education process, calculate education level, and enroll a person into an academy.
+ */
+public class EducationController {
+    /**
+     * Begins the education process for a Person in a Campaign.
+     *
+     * @param campaign The Campaign in which the education process is taking place.
+     * @param person The Person who is enrolling for education.
+     * @param academySet The set name of the academy.
+     * @param academyNameInSet The name of the academy within the set.
+     * @param courseIndex The index of the course in the academy.
+     * @param campus The campus location (can be null if the academy is local).
+     * @param faction The faction of the person.
+     */
+    public static void beginEducation(Campaign campaign, Person person, String academySet, String academyNameInSet, int courseIndex, String campus, String faction) {
+        ResourceBundle resources = ResourceBundle.getBundle("mekhq.resources.Education", MekHQ.getMHQOptions().getLocale());
+
+        Academy academy = getAcademy(academySet, academyNameInSet);
+
+        if (academy == null) {
+            LogManager.getLogger().error("No academy found with name {} in set {}", academyNameInSet, academySet);
+            return;
+        }
+
+        double tuition = academy.getTuitionAdjusted(person) * academy.getFactionDiscountAdjusted(campaign, person);
+
+        if (tuition > 0) {
+            // check there is enough money in the campaign & if so, make a debit
+            if (campaign.getFinances().getBalance().isLessThan(Money.of(tuition))) {
+                campaign.addReport(resources.getString("insufficientFunds.text").replaceAll("0", person.getHyperlinkedFullTitle()));
+                return;
+            } else {
+                campaign.getFinances().debit(TransactionType.EDUCATION, campaign.getLocalDate(), Money.of(tuition), resources.getString("payment.text")
+                        .replaceAll("0", person.getFullName()));
+            }
+        }
+
+        // with the checks done, and tuition paid, we can enroll Person
+        person.setEduCourseIndex(courseIndex);
+
+        enrollPerson(campaign, person, academy, campus, faction, courseIndex);
+
+        if (academy.isClan()) {
+            campaign.addReport(resources.getString("offToSchoolClan.text")
+                    .replaceAll("0", person.getFullName())
+                    .replaceAll("1", person.getEduAcademyName())
+                    .replaceAll("2", String.valueOf(person.getEduDaysOfTravelToAcademy())));
+        } else {
+            campaign.addReport(resources.getString("offToSchool.text")
+                    .replaceAll("0", person.getFullName())
+                    .replaceAll("1", person.getEduAcademyName())
+                    .replaceAll("2", String.valueOf(person.getEduDaysOfTravelToAcademy())));
+        }
```

Comment:
Adjust to use String interpolation instead. You're already making the localized text so will know what goes where and is more secure.

Reply:
Very easy fix. I didn't know about `String.format()` when writing this. I now do, and I absolutely agree that's a far better way of handling these.

It should have occurred to me to go back and use it here, too.